kubespawner/spawner.py

```diff
@@ -1809,18 +1836,32 @@ def _expand_user_properties(self, template):
         safe_username = escapism.escape(
             self.user.name, safe=safe_chars, escape_char='-'
         ).lower()
+
+        if self.slug_scheme == "safe":
+            # safe slug scheme escapes _after_ rendering the template
+            username = self.user.name
+            servername = raw_servername
+        else:
+            username = safe_username
+            servername = safe_servername
+
         rendered = template.format(
             userid=self.user.id,
-            username=safe_username,
+            username=username,
+            escaped_username=safe_username,
             unescaped_username=self.user.name,
             legacy_escape_username=legacy_escaped_username,
-            servername=safe_servername,
+            servername=servername,
             unescaped_servername=raw_servername,
+            escaped_servername=safe_servername,
```

Comment:
Same, do we really need this extra parameter ?

Reply:
same answer. Probably not, but this is the only way to preserve capabilities to produce the same strings.